### PR TITLE
Allow refType to be defined in @prop and @arrayProp

### PR DESCRIPTION
**Pull request by mfulop**
Merged changes from pr #149, solved conflicts, added itemsRefType (works the same as refType) to @arrayProps and tests

### CODE REVIEW

File: src/prop.ts [hunk 1]
```diff
@@ -369,22 +387,15 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
...
 
-/**
- * Set Property Options for the property below
- * @param options Options
- * @public
- */
-export function prop(options: PropOptionsWithValidate = {}) {
-  return (target: any, key: string) => {
-    const Type = (Reflect as any).getMetadata('design:type', target, key);
+export const prop = (options: PropOptionsWithValidate = {}) => (target: any, key: string) => {
```
Comment (hasezoey) [inline src/prop.ts]: please use the new style of functions

File: test/models/refTestBuffer.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import * as mongoose from 'mongoose';
+
+import { prop, Typegoose } from '../../src/typegoose';
+
+export class RefTestBuffer extends Typegoose {
```
Comment (hasezoey) [inline test/models/refTestBuffer.ts]: could you combine these into one file?

File: test/models/refTestNumber.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import * as mongoose from 'mongoose';
+
+import { prop, Typegoose } from '../../src/typegoose';
+
+export class RefTestNumber extends Typegoose {
```
Comment (hasezoey) [inline test/models/refTestNumber.ts]: could you combine these into one file?

File: test/models/refTestString.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import * as mongoose from 'mongoose';
+
+import { prop, Typegoose } from '../../src/typegoose';
+
+export class RefTestString extends Typegoose {
```
Comment (hasezoey) [inline test/models/refTestString.ts]: could you combine these into one file?

File: test/tests/ref.test.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    expect((RefTest.schema.path('refArrayBuffer2') as any).instance).to.equal('Array');
+    expect((RefTest.schema.path('refArrayBuffer2') as any).caster.instance).to.equal('Buffer');
+    expect((RefTest.schema.path('refArrayBuffer2') as any).caster.options.ref).to.equal('RefTestBuffer');
+  });
+
+  it('check reference with string _id', async () => {
+    const id1: string = 'testid1';
+    const id2: string = 'testid2';
+
+    const t = new RefTest();
```
Comment (hasezoey) [inline test/tests/ref.test.ts]: could you give it a proper variable name?

File: test/tests/ref.test.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    t.refFieldString = id1; // should not throw compilation error
```
Comment (hasezoey) [inline test/tests/ref.test.ts]: could you remove `// should not throw compilation error` on all lines?

**Review by hasezoey — CHANGES_REQUESTED** [review]
when these are resolved, could you squash some of your commits?

i would recommend reverting 00ddb93 and then rebase, not merge :)

**Comment by mfulop** [comment]
Sure will do. vscode merge somehow messed up things, will try another way, sorry

File: test/tests/ref.test.ts (hunk 6 above)
Comment (mfulop) [reply test/tests/ref.test.ts]: Yes. The idea was that these lines are actually test of the Ref declaration being correct. Otherwise test compilation would fail on these lines

**Comment by hasezoey** [comment]
i will review the changed in 60aa3ca tomorrow

**Comment by mfulop** [comment]
ok, you might want to look at ab91ee0 instead

**Comment by mfulop** [comment]
Btw though it’s working, I think Ref is now a bit too liberal. To be 100% correct, in Ref<R,T>, T should be R or the type of the designated id field of R.
Sadly it is beyond my current typescript knowledge :)

File: src/prop.ts [hunk 7]
```diff
@@ -58,6 +58,7 @@ export interface PropOptions extends BasePropOptions {
   ref?: any;
   /** Take the Path and try to resolve it to a Model */
   refPath?: string;
+  refType?: 'number' | 'string' | 'Buffer' | 'ObjectID';
```
Comment (hasezoey) [inline src/prop.ts]: change to lower case, just to keep everything normal
```suggestion
  refType?: 'number' | 'string' | 'buffer' | 'objectid;
```

File: src/prop.ts [hunk 8]
```diff
@@ -197,18 +198,35 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
     initAsObject(name, key);
   }
 
+  let refType;
+
+  switch (rawOptions.itemsRefType || rawOptions.refType) {
```
Comment (hasezoey) [inline src/prop.ts]: change to lowercase to keep everything normal, like above, and everything following
```suggestion
  switch ((rawOptions.itemsRefType || rawOptions.refType).toLowerCase()) {
```

File: src/prop.ts [hunk 9]
```diff
@@ -197,18 +198,35 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
...
+  case 'number':
+    refType = mongoose.Schema.Types.Number;
+    break;
+  case 'string':
+    refType = mongoose.Schema.Types.String;
+    break;
+  case 'Buffer':
```
Comment (hasezoey) [inline src/prop.ts]:
```suggestion
  case 'buffer':
```

File: src/prop.ts [hunk 10]
```diff
@@ -197,18 +198,35 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
...
+    refType = mongoose.Schema.Types.Buffer;
+    break;
+  default:
```
Comment (hasezoey) [inline src/prop.ts]:
```suggestion
  case 'objectid':
  default:
```

File: src/prop.ts [hunk 11]
```diff
@@ -395,7 +413,10 @@ export interface ArrayPropOptions extends BasePropOptions {
   itemsRef?: any;
   /** Same as {@link PropOptions.refPath}, only that it is for an array */
   itemsRefPath?: any;
+  /** Same as {@link PropOptions.refType}, only that it is for an array */
+  itemsRefType?: 'number' | 'string' | 'Buffer' | 'ObjectID';
```
Comment (hasezoey) [inline src/prop.ts]:
```suggestion
  itemsRefType?: 'number' | 'string' | 'buffer' | 'objectid';
```

**Review by hasezoey — CHANGES_REQUESTED** [review]
i would request setting the types to all lowercase to keep everything in style and not "string" and "Buffer" just to write it easy

PS: sorry that it took so long, my Desktop Environment just kept crashing

PPS: could you please squash all your commits? (`git rebase -i master`)
and try to reword your commits to comply with [CONTRIBUTING.md](https://github.com/szokodiakos/typegoose/blob/master/.github/CONTRIBUTING.md#how-to-structure-commits) (~~if not, i will at the merge squash & reword it~~)

**Comment by mfulop** [comment]
How about now? 
Another thing came to mind before we finalize this. I merged Matt's changes from #149 regarding the refType, but does that have to be string? Would not specifying actual schema types make more sense? (it would to me...):
`@prop({ ref: Car, refType: mongoose.Schema.Types.String })`

**Comment by hasezoey** [comment]
>  I merged Matt's changes from #149 regarding the refType, but does that have to be string?

yes, would be a better idea, only that "I merged Matt's changes" doesnt not seem to be actually merged (its still string)

**Comment by mfulop** [comment]
OK, so here c0ab190 is how I would do it. I think it's much cleaner this way. What do you think?

File: src/prop.ts [hunk 12]
```diff
@@ -201,14 +208,14 @@ function baseProp(rawOptions: any, Type: any, target: any, key: string, whatis:
   if (typeof ref === 'string') {
     schema[name][key] = {
       ...schema[name][key],
-      type: mongoose.Schema.Types.ObjectId,
+      type: rawOptions.refType || mongoose.Schema.Types.ObjectId,
```
Comment (hasezoey) [inline src/prop.ts]: i think it would be better to do something like `rawOptions.refType = rawOptions.refType ? rawOptions.refType : mongoose.Schema.Types.ObjectId` instead of repetitively doing this

**Review by hasezoey — CHANGES_REQUESTED** [review]

**Comment by hasezoey** [comment]
I noticed after some testing, that if i do 
```ts
const test = new RefTestModel({ refFieldString: new RefTestStringModel() }); // just as an example
if (isDocument(test.refFieldString)) {
  test.refFieldString // i mean here
}
```
the type of `test.refFieldString` in the if is `InstanceType<T> | (string & Document)` where it should only be `InstanceType<T>`

**Comment by hasezoey** [comment]
got it fixed, please modify typeguards.ts:
```ts
export function isDocument<T, S extends RefType>(doc: Ref<T, S>): doc is InstanceType<T> {
```
and
```ts
export function isDocumentArray<T, S extends RefType>(docs: Ref<T, S>[]): docs is InstanceType<T>[] {
```

**Comment by mfulop** [comment]
I implemented all the changes and added the test. That actually brought up this:
Shouldn't 
`export type Ref<R, T extends RefType = mongoose.Schema.Types.ObjectId> = R | T;`
be
`export type Ref<R, T extends RefType = mongoose.Types.ObjectId> = R | T;`
instead? (the actual values are not SchemaTypes but ObjectIds).

But would this change break compatibility?

**Comment by hasezoey** [comment]
a mongoose issue explaining the difference (for further readers) https://github.com/Automattic/mongoose/issues/1671

so yes, it would be *more* right to use `Types.T` insteace of `Schema.Types.T`
and i cant see any obvious "breaking" changes, "Ref<T, S>" is just a type and so-or-so it works as expected

so i would say, yes it would be good to keep it so

**Comment by mfulop** [comment]
This: `new RefTestModel({ refField: mongoose.Types.ObjectId() })` is actually not possible without casting when using SchemaType

File: test/models/refTests.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
+import * as mongoose from 'mongoose';
+
+import { arrayProp, prop, Ref, Typegoose } from '../../src/typegoose';
+
+export class RefTestBuffer extends Typegoose {
+  @prop() public _id: mongoose.Schema.Types.Buffer;
```
Comment (hasezoey) [inline test/models/refTests.ts]: it would be good if we could keep a consistent class style
so i ask, could you change this (and the following classes) to use the following style?
```ts
@prop()
public _id; // separate line
```

File: test/tests/ref.test.ts [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    expect((RefTestModel.schema.path('refArrayBuffer') as any).instance).to.equal('Array');
+    expect((RefTestModel.schema.path('refArrayBuffer') as any).caster.instance).to.equal('Buffer');
+    expect((RefTestModel.schema.path('refArrayBuffer') as any).caster.options.ref).to.equal('RefTestBuffer');
+    expect((RefTestModel.schema.path('refArrayBuffer2') as any).instance).to.equal('Array');
+    expect((RefTestModel.schema.path('refArrayBuffer2') as any).caster.instance).to.equal('Buffer');
+    expect((RefTestModel.schema.path('refArrayBuffer2') as any).caster.options.ref).to.equal('RefTestBuffer');
+  });
+
+  it('check reference with string _id', async () => {
+    const id1: string = 'testid1';
```
Comment (hasezoey) [inline test/tests/ref.test.ts]: to keep consistent style, could you remove the "redundant" typedef?

File: test/tests/ref.test.ts [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    const id2: string = 'testid2';
```
Comment (hasezoey) [inline test/tests/ref.test.ts]: to keep consistent style, could you remove the "redundant" typedef?

File: test/tests/ref.test.ts [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+    expect(isDocument(idFields.refFieldString)).to.equal(false);
+    expect(isDocument(idFields.refFieldNumber)).to.equal(false);
+    expect(isDocument(idFields.refFieldBuffer)).to.equal(false);
+
+    expect(isDocumentArray(idFields.refArray)).to.equal(false);
+    expect(isDocumentArray(idFields.refArrayString)).to.equal(false);
+    expect(isDocumentArray(idFields.refArrayNumber)).to.equal(false);
+    expect(isDocumentArray(idFields.refArrayBuffer)).to.equal(false);
+
+    const populatedFields = new RefTestModel({
```
Comment (hasezoey) [inline test/tests/ref.test.ts]: it would be good if this could be changed to store it and retrieve it, just to be "extra" sure
```suggestion
    const saved = await RefTestModel.create({
```

**Comment by hasezoey** [comment]
> This: `new RefTestModel({ refField: mongoose.Types.ObjectId() })` is actually not possible without casting when using SchemaType

ok, then lets keep it at `Schema.Types.T`

**Comment by mfulop** [comment]
Then its done ;-)

File: test/tests/ref.test.ts [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+      refField: new RefTestModel(),
+      refArray: [new RefTestModel(), new RefTestModel()],
+      refFieldString: new RefTestStringModel({ _id: 'test' }),
+      refArrayString: [new RefTestStringModel({ _id: 'test' }), new RefTestStringModel({ _id: 'test' })],
+      refFieldNumber: new RefTestNumberModel({ _id: 1234 }),
+      refArrayNumber: [new RefTestNumberModel({ _id: 1234 }), new RefTestNumberModel({ _id: 1234 })],
+      refFieldBuffer: new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+      refArrayBuffer: [new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) })]
+    });
```
Comment (hasezoey) [inline test/tests/ref.test.ts]: (in conjunction with the comment above)
```suggestion
    });
    
    const found = RefTestModel.findById(saved.id).populate('refField') //etc
```

PS: i know this is redundant, and will result in a long .populate "command", but just to be extra sure

File: test/tests/ref.test.ts [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+    const populatedFields = new RefTestModel({
+      refField: new RefTestModel(),
+      refArray: [new RefTestModel(), new RefTestModel()],
+      refFieldString: new RefTestStringModel({ _id: 'test' }),
+      refArrayString: [new RefTestStringModel({ _id: 'test' }), new RefTestStringModel({ _id: 'test' })],
+      refFieldNumber: new RefTestNumberModel({ _id: 1234 }),
+      refArrayNumber: [new RefTestNumberModel({ _id: 1234 }), new RefTestNumberModel({ _id: 1234 })],
+      refFieldBuffer: new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
+      refArrayBuffer: [new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
```
Comment (hasezoey) [inline test/tests/ref.test.ts]: (i suggest this change, because vscode on save somehow formats it wrong ...)
```suggestion
      refArrayBuffer: [
        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }),
```

File: test/tests/ref.test.ts [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) })]
```
Comment (hasezoey) [inline test/tests/ref.test.ts]:
```suggestion
        new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) })
      ]
```

**Review by hasezoey — CHANGES_REQUESTED** [review]
Everything is and looks good, just some small style changes

**Comment by mfulop** [comment]
Could you check something for me?

I am putting together the test you requested above. When running the below snippet, the created RefTestModel has both refFieldString and refFieldNumber fields created with the corresponding ID.
However, while the RefTestNumber document is created in the database, for some reason, the RefTestString document is not. Therefore, the next populate will not find it.
What makes the difference? Or is it a race condition somewhere?

```
await RefTestModel.create({
      refFieldString: new RefTestStringModel({ _id: 'test1' }), // document not created
      refFieldNumber: new RefTestNumberModel({ _id: 1234 }), // document created
    });
```

**Comment by mfulop** [comment]
By using create `refFieldString: await RefTestStringModel.create({ _id: 'test1' })` it works (of course) but I guess that's not the point.

**Comment by hasezoey** [comment]
> Or is it a race condition somewhere?

typegoose nowhere sets a race condition, at least not knowingly 

---

could you push it to another branch so i can check it out?

**Comment by mfulop** [comment]
https://github.com/mfulop/typegoose/tree/reftests/

**Comment by hasezoey** [comment]
> However, while the RefTestNumber document is created in the database, for some reason, the RefTestString document is not.

sorry, cant confirm that, for me everything gets saved
modified it to
```ts
const { _id: populatedId } = await RefTestModel.create({
  refField: new RefTestModel(), // gets saved
  refArray: [new RefTestModel(), new RefTestModel()], // gets saved
  refFieldString: new RefTestStringModel({ _id: 'test1' }), // gets saved
  refArrayString: [
    new RefTestStringModel({ _id: 'test2' }), // gets saved
    new RefTestStringModel({ _id: 'test3' }) // gets saved
  ],
  refFieldNumber: new RefTestNumberModel({ _id: 1234 }), // gets saved
  refArrayNumber: [new RefTestNumberModel({ _id: 5678 }), new RefTestNumberModel({ _id: 9876 })], // gets saved
  refFieldBuffer: new RefTestBufferModel({ _id: Buffer.from([1, 2, 3, 4]) }), // gets saved
  refArrayBuffer: [
    new RefTestBufferModel({ _id: Buffer.from([5, 6, 7, 8]) }), // gets saved
    new RefTestBufferModel({ _id: Buffer.from([9, 8, 7, 6]) }) // gets saved
  ]
});
```

**Comment by mfulop** [comment]
Do tests run without an error on the reftests branch for you?

**Comment by mfulop** [comment]
Oh yes, the root document is correctly created. But do the referenced documents (eg. the one with id 5678) all exist?

**Comment by hasezoey** [comment]
> Do tests run without an error on the reftests branch for you?

no thests do not run, because not populated properly

> Oh yes, the root document is correctly created. But do the referenced documents (eg. the one with id 5678) all exist?

i dont have any with the id with `5678`, only an array of numbers

**Comment by hasezoey** [comment]
currently it fails at `expect(foundPopulated.refArray.length).to.equal(2);`, with `AssertionError: expected 0 to equal 2`

**Comment by mfulop** [comment]
> currently it fails at `expect(foundPopulated.refArray.length).to.equal(2);`, with `AssertionError: expected 0 to equal 2`
yes and that means populate can't find the foreign document with that id

**Comment by hasezoey** [comment]
> yes and that means populate can't find the foreign document with that id

i can confirm that Field `refArrayNumber` documents are saved on the DB

**Comment by mfulop** [comment]
Ok, I will dig deeper tomorrow to see what’s going on. If the documents are there and the ids are there then populate should work.

**Comment by hasezoey** [comment]
got it further with
```ts
const pop = Object.keys(RefTestModel.schema.obj).join();
const foundPopulated = await RefTestModel.findById(populatedId).populate(pop);
```

new error: `AssertionError: expected false to equal true` on line
```ts
expect(isDocument(foundPopulated.refField)).to.equal(true);
```

will investigate further

**Comment by mfulop** [comment]
First thing I will try in the morning is the same with native mongoose.

**Comment by hasezoey** [comment]
Update(4):
i noticed after disabling all other tests, that the documents were never created (including the ones i said are existing), because they have duplicate ids (like the number `5678` is used 3 times)

TL;DR: what i said was wrong, no new refs are actually saved

**Comment by mfulop** [comment]
OK, so it is not supposed to work. Then await create() will do for tests, give me a minute

File: test/tests/ref.test.ts [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+        await RefTestNumberModel.create({ _id: 98765 })
+      ],
+      refFieldBuffer: await RefTestBufferModel.create({ _id: Buffer.from([1, 2, 3, 4, 5]) }),
+      refArrayBuffer: [
+        await RefTestBufferModel.create({ _id: Buffer.from([5, 6, 7, 8, 9]) }),
+        await RefTestBufferModel.create({ _id: Buffer.from([9, 8, 7, 6, 5]) })
+      ]
+    });
+
+    const populate = ['', 'String', 'Number', 'Buffer'].map((f) => `refField${f} refArray${f}`).join(' ');
```
Comment (hasezoey) [inline test/tests/ref.test.ts]:
```suggestion
    const populate = Object.keys(RefTestModel.schema.obj);
```

Comment (mfulop) [reply test/tests/ref.test.ts]: _id too?

Comment (hasezoey) [reply test/tests/ref.test.ts]: this dosnt include _id

Comment (hasezoey) [reply test/tests/ref.test.ts]:
```ts
const populate = Object.keys(RefTestModel.schema.obj);
// tslint:disable:no-
console.log('pop', populate);
```
outputs an array, without `_id`

**Comment by mfulop** [comment]
> -> we _could_ implement such things, but this would mean that we would need to redo the .create / .save functions OR do an "invisible" pre-save-hook
My 2 cents - I would not... it's a quite big difference compared to how mongoose behaves on its own, could be misleading for the ones with Mongoose experience

> --> but it is out of scope for _this_ PR
I agree

**Comment by hasezoey** [comment]
i see no problem anymore, only resolve the last review, squash your commits, and i will approve :+1:

**Comment by mfulop** [comment]
There you go 189a6ce

And thanks for all :)

**Review by hasezoey — APPROVED** [review]
Everything got resolved, everything conforms with the Coding Style, Commits are structured right :+1:

**Comment by hasezoey** [comment]
@Ben305 (if you are back) i would want to know if this should be in release 6.0.0 or release 5.10.0?

**Comment by mfulop** [comment]
Hi!

Sorry, I was out for a bit. Actually I would appreciate this before v6. I am planning to port some of our proprietary sw to Typegoose and this was one missing feature. (There might be more but then you’ll see a PR about them anyways :) ). Unfortunately this project is now iced for a few days but I’ll be back.

**Comment by hasezoey** [comment]
> Unfortunately this project is now iced for a few days but I’ll be back.

which project? this pr is finished an ready to go :) (ben is out for a week)

**Comment by mfulop** [comment]
The on where I am porting stuff to run on Typegoose :)

**Comment by hasezoey** [comment]
Merged it for now into 6.0.0(-19)

**Comment by hasezoey** [comment]
this can be closed because it is included in v6.0.0, and will not be included a v5.x release

@Ben305
